### PR TITLE
Added the Z algorithm.

### PR DESCRIPTION
**Pull request by Deeeemo**
This is what the algorithm does:
Given a string `s` of length `n`, calculate the array `int z[n]` where `z[i]` is the length of the longest substring starting at `i` that matches a prefix of `s`.
Runs in linear time.

Testing:
Used it to solve [this problem](https://codeforces.com/contest/1537/problem/E2).

### CODE REVIEW

**Comment by mattagar6** [comment]
We should put it into a function that returns a vector or something. It will be easier to use as a black box that way

**Comment by platers** [comment]
Agree it should be a function. I like the cp-algorithms implementation.
https://cp-algorithms.com/string/z-function.html#:~:text=time%20is%20linear.-,Implementation,-Implementation%20turns%20out

**Comment by Deeeemo** [comment]
Agreed, a function is better. I propose we use my version of Z, because I know how it works already.

File: Strings/Z.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#include<bits/stdc++.h>
+
+using namespace std;
+
+vector<int> get_z(string s) {
+    int n = s.size();
+    vector<int> z(n);
+    int l = 0, r = 0;
+    for (int i = 1; i < n; i++)
```
Comment (platers) [inline Strings/Z.cpp:9]: nit: Prefer curly brackets around multiline blocks for readability.

File: Strings/Z.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        if (r <= i){
+            l = r = i;
+            while (r < n && s[r] == s[r-l]) r++;
+            z[i] = r-i;
+        }
+        else{
+            if (z[i-l] < (r - i))
```
Comment (platers) [inline Strings/Z.cpp:16]: nit: prefer curly brackets here if the else block has them

File: Strings/Z.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            else{
+                l = i;
+                while (r < n && s[r] == s[r-l]) r++;
+                z[i] = r-i;
+            }
+        }
+    return z;
+}
+
+int main(){
```
Comment (platers) [inline Strings/Z.cpp:27]: I don't think a main is necessary. When its time to compile the pdf, its much easier to have no extra code in the file. I think we can get rid of the header as well.

**Review by mattagar6 — APPROVED** [review]
Looks good!

File: Strings/Z.cpp (hunk 1 above)
Comment (Deeeemo) [reply Strings/Z.cpp:9]: Nah

File: Strings/Z.cpp (hunk 3 above)
Comment (Deeeemo) [reply Strings/Z.cpp:27]: It's easier to test the code when there's a main. You can also use a main to show how to use the algo.

It's not a big deal for now and if we have a problem with fitting everything in the notebook, we'll deal with it then.

**Pull request merged**